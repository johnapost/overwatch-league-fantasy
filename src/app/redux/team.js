// @flow

import type { Player } from "../shared/player";

// Types
export type TeamState = {
  drafter: string | null,
  selectedPlayer: Player | null,
  roster: {
    [index: string]: Player | null
  }
};

type TeamDraftSelect = {
  type: "DRAFT_SELECT",
  player: Player
};

type TeamDraftPlace = {
  type: "DRAFT_PLACE",
  toIndex: number
};

type TeamMoveSelect = {
  type: "MOVE_SELECT",
  fromIndex: number
};

type TeamMovePlace = {
  type: "MOVE_PLACE",
  toIndex: number
};

type TeamSetDrafter = {
  type: "SET_DRAFTER",
  uid: number | null
};

type TeamAction =
  | TeamDraftSelect
  | TeamDraftPlace
  | TeamMoveSelect
  | TeamMovePlace
  | TeamSetDrafter;

// Action Creators
export const teamDraftSelect = (player: Player): TeamDraftSelect => ({
  type: "DRAFT_SELECT",
  player
});

export const teamDraftPlace = (toIndex: number): TeamDraftPlace => ({
  type: "DRAFT_PLACE",
  toIndex
});

export const teamMoveSelect = (fromIndex: number): TeamMoveSelect => ({
  type: "MOVE_SELECT",
  fromIndex
});

export const teamMovePlace = (toIndex: number): TeamMovePlace => ({
  type: "MOVE_PLACE",
  toIndex
});

export const teamSetDrafter = (uid: number | null): TeamSetDrafter => ({
  type: "SET_DRAFTER",
  uid
});

// Reducer
export const defaultState: TeamState = {
  drafter: null,
  selectedPlayer: null,
  roster: {}
};

export default (state: TeamState = defaultState, action: TeamAction) => {
  switch (action.type) {
    case "DRAFT_SELECT":
      return { ...state, selectedPlayer: action.player };
    case "DRAFT_PLACE":
      return {
        ...state,
        selectedPlayer: null,
        roster: {
          ...state.roster,
          [action.toIndex]: state.selectedPlayer
        }
      };
    case "MOVE_SELECT":
      return {
        ...state,
        selectedPlayer: state.roster[`${action.fromIndex}`]
      };
    case "MOVE_PLACE":
      return {
        ...state,
        selectedPlayer: null,
        roster: {
          ...state.roster,
          [Object.keys(state.roster).find(
            key => state.roster[key] === state.selectedPlayer
          )]: null,
          [action.toIndex]: state.selectedPlayer
        }
      };
    case "SET_DRAFTER":
      return {
        ...state,
        drafter: action.uid
      };
    default:
      return state;
  }
};
