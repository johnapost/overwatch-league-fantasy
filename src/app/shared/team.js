// @flow

import type { AbbreviatedName } from "../components/teamLogo";

export type Team = {
  abbreviatedName: AbbreviatedName,
  addressCountry: string,
  homeLocation: string,
  id: number,
  name: string,
  owl_division: number,
  primaryColor: string,
  secondaryColor: string
};
