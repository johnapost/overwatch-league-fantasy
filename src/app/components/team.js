// @flow

import React from "react";
import { compose } from "redux";
import { connect } from "react-redux";
import get from "lodash/get";
import withFirestore from "../shared/withFirestore";
import { teamDraftPlace } from "../redux/team";
import Header from "./header";
import Player from "./player";
import DraftSlot from "./draftSlot";

import type { TeamState } from "../redux/team";
import type { Player as PlayerType } from "../shared/player";
import type { Role } from "../shared/roles";
import type { Roster } from "../shared/roster";

type RosterSlots = {
  [index: string]: Role
};

type Props = TeamState & {
  drafting: boolean,
  roster: Roster,
  rosterSlots: RosterSlots,
  draftPlace: typeof teamDraftPlace
};

// Merge roster with rosterSlots to mix displays for draft
const mergeDrafted = (
  roster: Roster,
  rosterSlots: RosterSlots
): (Role | PlayerType)[] =>
  Object.entries(rosterSlots).reduce(
    (accum, [key, value]) => [
      ...accum,
      // Null roster values are replaced with DraftSlot component
      roster[key] ? { ...roster[key] } : ((value: any): PlayerType)
    ],
    []
  );

export const Team = ({
  drafting,
  draftPlace,
  roster,
  rosterSlots,
  selectedPlayer
}: Props) => {
  // Creates click handler for each component
  const createOnClick = (index: number) => (role: Role) => (e: Event) => {
    e.preventDefault();
    // Guard against out-of-order drafting
    if (!drafting) return;
    // Guard against empty placements
    if (!selectedPlayer) return;
    // Guard against duplicates
    if (
      Object.values(roster).some(
        player =>
          player &&
          player.id &&
          selectedPlayer &&
          player.id === selectedPlayer.id
      )
    )
      return;
    if (
      // Guard against off-role picks
      selectedPlayer &&
      selectedPlayer.attributes.role !== role &&
      // But, flex roles can go wherever they want
      role !== "flex" &&
      selectedPlayer.attributes.role !== "flex"
    )
      return;
    draftPlace(index);
  };

  return (
    <div>
      <Header title="My Team" />
      <div className="slots-wrapper">
        {mergeDrafted(roster, rosterSlots).map((value, index) =>
          typeof value === "string" ? (
            <DraftSlot
              key={`${index + 1}`}
              role={value}
              onClick={createOnClick(index)(value)}
            />
          ) : (
            <Player
              key={value.id}
              onClick={createOnClick(index)(value.attributes.role)}
              {...value}
            />
          )
        )}
      </div>
      <style jsx>{`
        .slots-wrapper {
          display: flex;
          flex-direction: row;
          flex-wrap: wrap;
          justify-content: center;
          margin: 25px;
        }
      `}</style>
    </div>
  );
};

export const mapStateToProps = ({ firestore, team, user: { uid } }) => {
  const rosterSlots = get(firestore.data, "leagues.first.rosterSlots", {});
  const drafter = get(firestore.data, "leagues.first.drafter", "");

  return {
    ...team,
    drafting: drafter === uid,
    rosterSlots
  };
};

const mapDispatchProps = {
  draftPlace: teamDraftPlace
};

export default compose(
  connect(
    mapStateToProps,
    mapDispatchProps
  ),
  withFirestore(() => [
    {
      collection: "leagues",
      doc: "first"
    }
  ])
)(Team);
