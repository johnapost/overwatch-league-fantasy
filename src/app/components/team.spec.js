import React from "react";
import { shallow } from "enzyme";
import { TeamComponent, mapStateToProps } from "./team";
import DraftSlot from "./draftSlot";
import Player from "./player";

const props = {
  drafting: false,
  roster: {
    0: {
      attributes: {
        role: "offense"
      },
      headshot: "headshot.jpg",
      id: "999",
      name: "Nina",
      teamId: 4523,
      onClick: jest.fn(),
      role: undefined
    }
  },
  rosterSlots: {
    0: "offense",
    1: "tank"
  },
  draftPlace: () => {}
};

describe("Team", () => {
  describe("component", () => {
    const wrapper = shallow(<TeamComponent {...props} />);

    it("should render a DraftSlot component", () => {
      expect(wrapper.find(DraftSlot).length).toBe(1);
    });

    it("should render a Player component", () => {
      expect(wrapper.find(Player).length).toBe(1);
    });
  });

  describe("mapStateToProps", () => {
    const store = {
      firestore: {
        data: {
          leagues: {
            first: {
              drafter: "123"
            }
          }
        }
      },
      team: {
        drafter: null,
        selectedPlayer: null,
        roster: {}
      },
      user: {
        uid: "123"
      }
    };

    it("should check if currently drafting", () => {
      expect(mapStateToProps(store)).toMatchSnapshot();
    });

    it("should assign rosterSlots", () => {
      expect(
        mapStateToProps({
          ...store,
          rosterSlots: {
            0: "offense",
            1: "tank"
          }
        })
      ).toMatchSnapshot();
    });
  });
});
