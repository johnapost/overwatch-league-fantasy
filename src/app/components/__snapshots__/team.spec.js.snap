// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Team mapStateToProps should assign rosterSlots 1`] = `
Object {
  "drafter": null,
  "drafting": true,
  "roster": Object {},
  "rosterSlots": Object {},
  "selectedPlayer": null,
}
`;

exports[`Team mapStateToProps should check if currently drafting 1`] = `
Object {
  "drafter": null,
  "drafting": true,
  "roster": Object {},
  "rosterSlots": Object {},
  "selectedPlayer": null,
}
`;
